doc/mgmt/gnmi/master_arbitration.md

```diff
@@ -0,0 +1,331 @@
+// is disabled.
+func ReqFromMasterDisabledMA(req *gnmipb.SetRequest) bool {
+  return true
+}
+
+// ReqFromMaster point to a function that is called to verify if the request comes
+// from a master controller.
+ReqFromMaster := ReqFromMasterDisabledMA
+
+func (s *Server) Set(ctx context.Context, req *gnmipb.SetRequest) (*gnmipb.SetResponse, error) {
+  if !reqFromMaster(req) {
+    return nil, status.Error(codes.PermissionDenied, "Not a master")
+  }
+
+  // ...
+
+```
+
+If Master Arbitration is enabled, the following check is executed:
+
+```
+type uint128 struct {
+  High uint64
+  Low uint64
+}
+
+func (lh *uint128) Compare(rh *uint128) {
+  if lh.High > rh.High {
+    return 1;
+  }
+  if lh.High < rh.High {
+    return -1;
+  }
+  if lh.Low > rh.Low {
+    return 1;
+  }
+  if lh.Low < rh.Low {
+    return -1;
+  }
+  return 0;
+}
+
+var masterEID = uint128{High: 0, Low: 0}
+
+// ReqFromMasterEnabledMA returns true if the request is sent by the master
+// controller.
+func ReqFromMasterEnabledMA(req *gnmipb.SetRequest) bool {
+  // Read the election_id.
+  reqEID := uint128{High: 0, Low: 0}
+  // It can be one of many extensions, so iterate through them to find it.
+  for _, e := range req.extension {
+    ma := e.GetMasterArbitration()
+    if ma == nil {
+      continue
+    }
+    // The Master Arbitration descriptor has been found.
+    reqEID = uint128{High: ma.ElectionID.High, Low = ma.ElectionID.Low}
+    // Use the election ID that is in the last extension, so, no 'break' here.
+  }
+  switch masterEID.Compare(reqEID) {
+  case 1:  // This Election ID is smaller than the known Master Election ID.
+    return false
+  case -1: // New Master Election ID received!
+    masterEID = reqEID
+  }
+  return true
+}
+```
+
+#### Manifest (if the feature is an Application Extension)
+
+N/A
+
+#### CLI/YANG model Enhancements
+
+N/A
+
+#### Config DB Enhancements
+
+N/A
+
+### Warmboot and Fastboot Design Impact
+
+N/A
+
+### Restrictions/Limitations
+
+Master Arbitration is designed to enable one master client for one role. However, only default role is supported for now.  In other words, the content of the role field is ignored.
+
+### Testing Requirements/Design
+The main focus of the test is on the coverage of the new feature. The user cases are listed as following:
+
+* No Master Arbitration extension protobuf when Master Arbitration is disabled
+  * Nothing should happen
+
+* No Master Arbitration extension protobuf when Master Arbitration is enabled
+  * The EID = 0 should be used
+
+* Multiple arbitration extension protobufs when Master Arbitration is disabled
+  * Nothing should happen
+
+* Multiple Master Arbitration extension protobufs when Master Arbitration is enabled
+  * The last one should be used.
+
+* Change of the master taking place
+  * Accepted call from EID = 1
+  * Accepted call from EID = 2
+  * Rejected call from EID = 1
+
+* Restart of the switch resets the expected master EID to 0
```

Comment:
If two controllers send conflicting requests with the same largest EID=2, then one config will overwrite the other. That appears to be concerning. 
Also, I assume all controller instances will exchange the same certificate for authentication. Please confirm.

Reply:
> If two controllers send conflicting requests with the same largest EID=2, then one config will overwrite the other. That appears to be concerning.
Yes. This concerning is EXACTLY the one when Master Arbitration is not available at all. The discussion of implementation of EID is out of the scope of switch implementation though. 

> Also, I assume all controller instances will exchange the same certificate for authentication. Please confirm.
Each controller (no matter it is selected as Master or not) has its own authentication. The exchange is not necessary.

Reply:
>> Restart of the switch resets the expected master EID to 0
If we reset EID to 0 upon switch reset, wont we accept the first set request with lower EID > 0 from the client?

Reply:
It is possible. As long as the controller manager be able to assign the highest EID to the previous master controller, the previous master controller will take control eventually. Comparing the drawback of possible unreachable switch after EID rollover vs loose of master controller for a short time period, we think the later one is more acceptable.